pysph/cpy/jit.py

```diff
@@ -0,0 +1,523 @@
+from functools import wraps
+from textwrap import dedent, wrap
+
+from mako.template import Template
+import numpy as np
+import inspect
+import ast
+import importlib
+import warnings
+import sys
+
+from .config import get_config
+from .cython_generator import get_parallel_range, CythonGenerator
+from .transpiler import (
+    Transpiler,
+    convert_to_float_if_needed,
+    filter_calls,
+    get_external_symbols_and_calls,
+    BUILTINS)
+from .types import (dtype_to_ctype, get_declare_info,
+                    dtype_to_knowntype)
+from .extern import Extern
+from .ast_utils import get_unknown_names_and_calls
+
+import pysph.cpy.array as array
+import pysph.cpy.parallel as parallel
+import pysph.cpy.types as types
+
+
+def annotate(func):
+    def wrapper(func):
+        func.is_jit = True
+        return func
+
+    if func is None:
+        return wrapper
+    else:
+        return wrapper(func)
+
+
+def memoize(f):
+    @wraps(f)
+    def wrapper(obj, *args, **kwargs):
+        if not hasattr(obj, 'cache'):
+            obj.cache = dict()
+        key = tuple([get_ctype_from_arg(arg) for arg in args])
+        if key not in obj.cache:
+            obj.cache[key] = f(obj, *args, **kwargs)
+        return obj.cache[key]
+    return wrapper
+
+
+def getargspec(f):
+    if sys.version_info[0] < 3:
+        return inspect.getargspec(f)[0]
+    else:
+        return inspect.getfullargspec(f)[0]
+
+
+def get_ctype_from_arg(arg):
+    if isinstance(arg, array.Array):
+        return arg.gptr_type
+    elif isinstance(arg, np.ndarray):
+        return dtype_to_ctype(arg.dtype)
+    else:
+        if isinstance(arg, float):
+            return 'double'
+        else:
+            return 'long'
+
+
+def get_binop_return_type(a, b):
+    if a is None or b is None:
+        return None
+    preference_order = ['short', 'int', 'long', 'float', 'double']
+    unsigned_a = unsigned_b = False
+    if a.startswith('u'):
+        unsigned_a = True
+        a = a[1:]
+    if b.startswith('u'):
+        unsigned_b = True
+        b = b[1:]
+    idx_a = preference_order.index(a)
+    idx_b = preference_order.index(b)
+    return_type = preference_order[idx_a] if idx_a > idx_b else \
+        preference_order[idx_b]
+    if unsigned_a and unsigned_b:
+        return_type = 'u%s' % return_type
+    return return_type
+
+
+class AnnotationHelper(ast.NodeVisitor):
+    def __init__(self, func, arg_types):
+        self.func = func
+        self.arg_types = arg_types
+        self.var_types = arg_types.copy()
+        self.children = {}
```

Comment:
Can you add a comment on what "children" means here or perhaps a better variable name?